test/controllers/glossary_controller_test.rb

```diff
@@ -191,3 +191,46 @@ def update_and_reload_plane_past_version
     assert_template(:show_past_glossary_term, partial: "_glossary_term")
   end
 end
+
+# tests of image handling
+class GlossaryControllerImageTest < GlossaryControllerTest
+  def setup
+    @controller = GlossaryController.new
+  end
+
+  def term_with_image_params
+    setup_image_dirs
+    file = "#{::Rails.root}/test/images/Coprinus_comatus.jpg"
+    file = Rack::Test::UploadedFile.new(file, "image/jpeg")
+
+    {
+      glossary_term:  { name: "Pancake", description: "Flat" },
+      copyright_holder: "Me",
+      date: { copyright_year: 2013 },
+      upload: { license_id: licenses(:ccnc25).id },
+      image: {
+        "0" => {
+          image: file,
+          copyright_holder: "zuul",
+          when: Time.current
+        }
+      }
+    }
+  end
+
+  def test_process_image_failure
+    login("rolf")
+    # Simulate process_image failure.
+    Image.any_instance.stubs(:process_image).returns(false)
+    post(:create_glossary_term, term_with_image_params)
+    assert_flash_error
+  end
+
+  def test_image_save_failure
+    login("rolf")
+    # Simulate image.save failure.
+    Image.any_instance.stubs(:save).returns(false)
+    post(:create_glossary_term, term_with_image_params)
+    assert_empty(GlossaryTerm.last.images)
+  end
```

Comment:
I added these tests at the first commit in order to get 100% coverage of the controller.
**Questions**:
- Is 100% coverage a fetish? (Especially where the only uncovered lines are failures, see https://www.pivotaltracker.com/story/show/174310150/comments/217129032).
- Are there better tests for the uncovered lines, e.g., a way to test them without stubbing?
- Are there better/more meaningful assertions? (Especially for  `test_image_save_failure`)

Reply:
100% coverage is definitely not "a fetish".  Often it is considered the bare minimum for people who take testing seriously and it should not be that hard to get or maintain once you achieve it.  That being said, it is often harder to get there on legacy projects that weren't created with testing in mind.  MO falls into that category, but it is actually in way better shape that it was and it's in way better shape than most of the legacy projects I've worked on.  I have worked on systems that have 100% coverage and it makes a HUGE difference.  Essentially you never have to dig through coverage reports to figure out if the change you made is covered or not.  You just check that there is still 100% coverage.  If there is you move on.  The big win is when you want to make larger changes, you can really rely on the test suite to tell you if you broke something.

If you want to see a project that really takes testing seriously, check out the SQLite (https://www.sqlite.org/testing.html).  It's the most impressive one that I've seen.  I would guess we have roughly the same amount of code in our tests as in our code base.  SQLite has 640 TIMES as much test code as code in their library.  It not only has 100% code coverage, it has 100% BRANCH coverage (meaning every possible code path is tested. So for example when you do `foo ? bar : baz` it makes sure that there are cases where foo is true and false not just that the line was run at least once by the test suite.  There are much more complicated cases, but suffice to say 100% coverage is way easier to achieve than 100% branch coverage.  Beyond that (which SQLite also does) there is disaster and bad input testing.  They deal with testing for data file corruption, bad input, unstable hardware, all kinds of stuff.  Of course they are a key piece of infrastructure for a lot of really important applications so it makes sense that they look at stuff like that, but it helps give some perspective on what testing can be like.

Regarding stubbing, there are different theories here.  Some people do a lot of stubbing especially for unit tests.  Personally, when you don't have 100% coverage on a legacy project I avoid them because hopefully you can pick up some stuff you didn't have covered previously and it helps you exercise the sub-modules.  For controller/request/integration specs, I try to avoid them even more since there is something nice with ensure that all lines of code can be reached from a known request.  However, in this case you are essentially stubbing out the file system which is totally fine since if that doesn't work, then we're totally screwed and managing file system state in tests is hard.

In my view, your current test `test_image_save_failure` is very reasonable.  It's really clear what it is testing (failed saves) and what the expectations are (this doesn't break things and doesn't create add an image).  The only problem I see is using `.last` is a little bit risky, but unless the POST returns the id of the GlossaryTerm, I don't see a better way to do this.  I've worked with systems where the id is in the POST results, but I don't know off hand if that's true in MO.  Moving more towards standard CRUD may actually enable that.

Reply:
@mo-nathan: Thanks for the extended response.

**Branch Coverage**. I'll try to be more careful about using the ternary operator. I hadn't thought about its sometimes causing misleading coverage info. Maybe I shouldn't use it at all. (I tend to use it when it fits on one line; that reduces the number of lines in the method, class, and file.) 
**Are there other Ruby or Rails idioms which cause similar problems?**

**`.last`**. Our tests do this a lot. I hadn't thought about the fact that the last object returned from the db might not be the last one created. 
Here I could derive the id from `@response`-- knowing that `create` sends us to the `show` page for the created glossary term:
```ruby
id = @response.location.sub(%r{/#{glossary_terms_url}}, "")
```
But I don't like the extra dependency.
**What do you think?**

Reply:
@mo-nathan: What is the risk with `.last`? 

I was wrong to be thinking about possible random ordering from the db because Rails takes care of that. 
For `GlossaryTerm.last`, Rails generates: 
```sql 
SELECT  `glossary_terms`.* FROM `glossary_terms` ORDER BY `glossary_terms`.`id` DESC LIMIT 1
```
@pellaea was right on about this. See https://github.com/MushroomObserver/mushroom-observer/pull/640#issuecomment-683504272

Reply:
So two scenarios I can think of where `.last` can go wrong are when tests are running in parallel and if we were to ever switch to another database since some databases (SQLite for one) actually reuse ids that have been deleted, so in this case that could fail intermittently.  I'm certainly aware that we use .last a lot in our tests (heck I was probably the first one to write such a test).  However, as my experience with writing tests has gotten deeper I've realized that it can be problematic.  Personally at this point I would far rather parse the id out the response as you propose than rely on the `.last` element happening to be the right one.  Practically speaking it shouldn't matter unless we go to running tests in parallel, but I thought I heard some potential interest in that recently.  However, I can also buy the argument that we should just keep doing it the way we are doing it.  At this point I would not recommend changing it for this PR.  As I said it's just "a little bit risky".

Reply:
Testing branch coverage is not somewhere I've personally gone, but I know there are test analysis tools that do that.  The ?: example is just a simple example that I thought would make it clear.  I would not sweat using that construct too much (although personally I tend to avoid it).  I'd far rather get to 100% coverage first!

As I understand, true branch coverage means you are aiming to test every possible path through the code.  A more complicated example of branch coverage are cases where there are two conditionals in sequence.  E.g.,
```
if apple
  banana
else
  cherry
end
if donut
  easter_egg
else
  flapjack
end
```
As I understand it to do true branch coverage you should test this with all 4 possible patterns of apple and banana.  However, coverage can be achieved with only two patterns (e.g., true/true and false/false).  Obviously this can lead to some serious exponential blowup if you have a lot of conditionals.  This is when unit testing (and small functions) becomes so important vs. request/integration testing.  As long as you've been through those conditionals in a unit test, it's much less important to test them from the perspective of higher level functions that call the unit tested functions.  However, I'm sure you could create cases that would require that.  Exactly how folks who do this in practice make this tractable I don't know, but I'm sure there are conventions for how to it well.